influxdb3_processing_engine/src/plugins.rs

```diff
@@ -208,25 +286,109 @@ mod python_plugin {
-                        .unwrap();
-                    self.write_buffer
-                        .write_lp(
-                            NamespaceName::new(db_name).unwrap(),
-                            output_lines.join("\n").as_str(),
-                            Time::from_timestamp_nanos(ingest_time.as_nanos() as i64),
-                            false,
-                            Precision::Nanosecond,
-                        )
-                        .await?;
-                }
+                db_writes.execute(&self.write_buffer).await?;
             }
 
             Ok(false)
         }
     }
+
+    struct DatabaseWriteBuffer {
+        writes: HashMap<String, Vec<String>>,
+    }
+    impl DatabaseWriteBuffer {
+        fn new() -> Self {
+            Self {
+                writes: HashMap::new(),
+            }
+        }
+
+        fn add_lines(&mut self, db_name: &str, lines: Vec<String>) {
+            self.writes.entry_ref(db_name).or_default().extend(lines);
+        }
+
+        fn is_empty(&self) -> bool {
+            self.writes.is_empty()
+        }
+
+        async fn execute(self, write_buffer: &Arc<dyn WriteBuffer>) -> Result<(), Error> {
+            for (db_name, output_lines) in self.writes {
+                let ingest_time = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap();
+                write_buffer
+                    .write_lp(
+                        NamespaceName::new(db_name).unwrap(),
+                        output_lines.join("\n").as_str(),
+                        Time::from_timestamp_nanos(ingest_time.as_nanos() as i64),
+                        false,
+                        Precision::Nanosecond,
+                    )
+                    .await?;
+            }
+            Ok(())
+        }
+    }
+
+    struct CronTriggerRunner {
+        schedule: OwnedScheduleIterator<Utc>,
+        next_trigger_time: Option<DateTime<Utc>>,
+    }
+    impl CronTriggerRunner {
+        fn new(cron_schedule: Schedule) -> Self {
+            let mut schedule = cron_schedule.upcoming_owned(Utc);
+            let next_trigger_time = schedule.next();
+            Self {
+                schedule,
+                next_trigger_time,
+            }
+        }
+
+        async fn run_at_time(
+            &mut self,
+            plugin: &TriggerPlugin,
+            db_schema: Arc<DatabaseSchema>,
+        ) -> Result<(), Error> {
+            let Some(trigger_time) = self.next_trigger_time else {
+                return Err(anyhow!("running a cron trigger that is finished.").into());
+            };
+            let result = execute_cron_trigger(
+                &plugin.plugin_code,
+                trigger_time,
+                Arc::clone(&db_schema),
+                Arc::clone(&plugin.query_executor),
+                &plugin.trigger_definition.trigger_arguments,
+            )
+            .expect("should be able to run trigger");
```

Comment:
any reason this might error, can we catch this and log it?

Reply:
Yeah, any plugin error would trigger, switching to just have it exit with `?`.